src/rust/engine/graph/src/entry.rs

```diff
@@ -432,13 +440,14 @@ impl<N: Node> Entry<N> {
           //
           // On the other hand, if the Node is uncacheable, we store the previous result as
           // Uncacheable, which allows its value to be used only within the current Run.
+          let item = result.get_item();
           Self::run(
             context,
             &self.node,
             entry_id,
             run_token,
             generation,
-            if self.node.cacheable() {
+            if self.node.cacheable(Some(item)) {
```

Comment:
@stuhood using `peek()` here (see https://github.com/gshuflin/pants/commit/5fd85b538f2b09d326cd91d17cdd9ea44ea0aff6) results in the correct behavior when actually running a failing process in pants, but causes three `graph` tests to fail:
```failures:

---- tests::invalidate_and_rerun stdout ----
thread 'tests::invalidate_and_rerun' panicked at 'assertion failed: `(left == right)`
  left: `[TNode(2, true), TNode(1, true)]`,
 right: `[TNode(1, true), TNode(2, true)]`', graph/src/tests.rs:92:3

---- tests::invalidate_and_clean stdout ----
thread 'tests::invalidate_and_clean' panicked at 'assertion failed: `(left == right)`
  left: `[TNode(2, true), TNode(1, true), TNode(0, true), TNode(2, true), TNode(1, true)]`,
 right: `[TNode(2, true), TNode(1, true), TNode(0, true), TNode(1, true)]`', graph/src/tests.rs:55:3

---- tests::dirty_dependents_of_uncacheable_node stdout ----
thread 'tests::dirty_dependents_of_uncacheable_node' panicked at 'assertion failed: `(left == right)`
  left: `[TNode(2, true), TNode(1, true), TNode(0, false)]`,
 right: `[TNode(0, true)]`', graph/src/tests.rs:297:3
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
```

I'm not sure if we should expect these failures with this change, and change the tests, or if we want the behavior and can't use `peek` in this way.

Reply:
The `graph` crate tests don't observe the value in your current PR, so it doesn't seem like it could cause those tests to fail? ie, the method is implemented as:
```
  fn cacheable(&self, _: Option<&Vec<T>>) -> bool {
    self.1
  }
```